special treatment and warning message for HIL platform arming

kd0aij opened the pull request:
replace preflight check failures with reminder to connect to simulator
Unfortunately, the message doesn't show up in the hilsim panel of QGC.

LorenzMeier commented:
It does. Look for the red icon in the toolbar.

LorenzMeier commented:
@dogmaphobic @DonLakeFlyer I think we need to do something about the string messages to the user. Essentially all of them are missing the notifications in the toolbar, even if they are experiencing issues. We probably need to display the critical ones as banner.
Essentially all my user testing is confirming this. Severin already filed an issue: https://github.com/mavlink/qgroundcontrol/issues/1586

LorenzMeier commented on src/modules/commander/commander.cpp:
```diff
@@ -430,6 +430,12 @@ transition_result_t arm_disarm(bool arm, const int mavlink_fd_local, const char
 	transition_result_t arming_res = TRANSITION_NOT_CHANGED;
 
+	// For HIL platforms, require that simulated sensors are connected
+	if (autostart_id < 2000 && status.hil_state != vehicle_status_s::HIL_STATE_ON) {
```
can we do the 2000 here as a define?

DonLakeFlyer commented:
This issue for me in using the dropdown is that the firmware is so noisy with messaging (both critical which aren't really critical, and non-critical) I never pay any attention to it. If you change critical to banner it will likely drive everyone crazy. For example when disarmed you continuously get critical messages about detected landing or something like that.

I believe In Mission Planner critical messages show up right over the HUD. So you know you shouldn't fly.

LorenzMeier commented:
I think those are two orthogonal issues: One is the message rate of the firmware, the other is the visibility of those. Just because its too noisy right now is not a reason to not use a banner. E.g. the takeoff is addressed on the release branch.

DonLakeFlyer commented:
> I think those are two orthogonal issues

Not really. If you make critical messaging more prominent, and the firmware still spits out too many critical messages which are not really relevant, one of two things will happen:
1) People will initially complain that it is too annoying
2) People will continue to not pay attention since there are too many they don't care about and/or the important one will pass by and be replaced with an unimportant one

LorenzMeier commented:
Yes, I totally agree. What I was trying to say is that we need to fix both, because unless both is fixed the user experience is horrible when the system runs into issues.

LorenzMeier commented on src/modules/commander/commander.cpp:
```diff
@@ -430,6 +433,13 @@ transition_result_t arm_disarm(bool arm, const int mavlink_fd_local, const char
 	transition_result_t arming_res = TRANSITION_NOT_CHANGED;
 
+	// For HIL platforms, require that simulated sensors are connected
+	if (autostart_id >= HIL_ID_MIN && autostart_id <= HIL_ID_MAX
```
This check looks like it could be a function ```is_hil_setup()```, as its used two times in this patch: ```autostart_id >= HIL_ID_MIN && autostart_id <= HIL_ID_MAX```.

LorenzMeier commented on src/modules/commander/commander.cpp:
```diff
@@ -430,6 +433,13 @@ transition_result_t arm_disarm(bool arm, const int mavlink_fd_local, const char
+	// For HIL platforms, require that simulated sensors are connected
+	if (autostart_id >= HIL_ID_MIN && autostart_id <= HIL_ID_MAX
+			&& status.hil_state != vehicle_status_s::HIL_STATE_ON) {
+		mavlink_and_console_log_critical(mavlink_fd_local, "HIL platform: Connect to simulator before arming");
```
It looks like the function should also check if the user is trying to go from disarm -> arm

LorenzMeier commented on src/modules/commander/commander.cpp:
```diff
@@ -1144,7 +1154,11 @@ int commander_thread_main(int argc, char *argv[])
 	// Run preflight check
 	param_get(_param_autostart_id, &autostart_id);
-	if (autostart_id > 1999) {
+	if (autostart_id >= HIL_ID_MIN && autostart_id <= HIL_ID_MAX) {
```
Same boilerplate code.

LorenzMeier commented on src/modules/commander/commander.cpp:
```diff
@@ -1327,14 +1336,14 @@ int commander_thread_main(int argc, char *argv[])
-						(void)Commander::preflightCheck(mavlink_fd, true, true, true, true, chAirspeed,
-								!(status.rc_input_mode == vehicle_status_s::RC_IN_MODE_OFF), !status.circuit_breaker_engaged_gpsfailure_check);
-					} else {
+					if (autostart_id >= HIL_ID_MIN && autostart_id <= HIL_ID_MAX) {
```
Same boilerplate code.

LorenzMeier commented:
@kd0aij Thanks! Only minor comments, but it would be great to have them addressed.

kd0aij commented:
when I tested this today, RTL didn't loiter correctly above home (flew at about 20' AGL, evenually hit the ground), and AUTO:mission dove straight into the ground, as though the waypoint altitudes were too low, but I hadn't changed the waypoints since the last good waypoints test.

LorenzMeier commented:
Do you have a log file of this?

kd0aij commented:
https://copy.com/0nTorqbF18iVifCZ
one log each for auto:mission and RTL

LorenzMeier commented:
The altitude setpoint was way too low:
http://dash.oznet.ch/view/CQJ8NqLQXKPA6V792mRs36#TECS_Outer_Loop_PLOT

kd0aij commented:
Any advice on root cause?

LorenzMeier commented:
Please check the MAVLink messages and ensure that the home position was set correctly in terms of altitude.

kd0aij commented:
rebased and added is_hil_setup()

changed Xplane airport location, cleared and set new waypoints; altitude now behaving correctly.
RTL flew at 42m above home. On reaching home location announced RTL: landing and continued flying forever at 40% throttle. Resetting Xplane to home position resulted in mavlink message RTL: completed, landed.
Will try changing back to the location which had the altitude problem.

kd0aij commented:
after changing to new airport and observing similar altitude problems, restarted QGC and pixhawk with Xplane running at the new airport. Created new waypoints and flew mission OK with loiter at last WP. Turned off TX and RTL worked properly with a normal landing sequence.

I guess the home location isn't being updated in the autopilot even when pressing the "set home" button in QGC, or else something else is stale.

LorenzMeier commented:
The set home command looks good - did the vehicle report the new position being set in the console?

LorenzMeier commented:
This one here will also be of interest:
https://github.com/PX4/Firmware/pull/2322

LorenzMeier commented:
And filed issue for RTL landing here:
https://github.com/PX4/Firmware/issues/2323

kd0aij commented:
I didn't see the nsh console message for home loc. when pressing the QGC set home button

LorenzMeier commented:
It took me a while to figure out which one you meant. I've removed it, since it wasn't connected to anything.

LorenzMeier commented:
Right-click onto the map and select "put home location here" to set the home location.

kd0aij commented:
Buttons that do nothing is an entertaining way to add confusion :)
Seriously, one thing I really don't like with the current trend in GUIs is to make everything so "intuitive" that the only way to find out how it works is to start "gesturing" at things to see what happens.
